sklearn/utils/stats.py

```diff
@@ -32,41 +32,60 @@ def _weighted_percentile(array, sample_weight, percentile=50):
     percentile : int if `array` 1D, ndarray if `array` 2D
         Weighted percentile.
     """
+    xp, _ = get_namespace(array)
     n_dim = array.ndim
     if n_dim == 0:
         return array[()]
     if array.ndim == 1:
-        array = array.reshape((-1, 1))
+        array = xp.reshape(array, (-1, 1))
     # When sample_weight 1D, repeat for each array.shape[1]
     if array.shape != sample_weight.shape and array.shape[0] == sample_weight.shape[0]:
-        sample_weight = np.tile(sample_weight, (array.shape[1], 1)).T
-    sorted_idx = np.argsort(array, axis=0)
-    sorted_weights = np.take_along_axis(sample_weight, sorted_idx, axis=0)
+        sample_weight = xp.tile(sample_weight, (array.shape[1], 1)).T
+    sorted_idx = xp.argsort(array, axis=0)
+    sorted_weights = _take_along_axis(sample_weight, sorted_idx, xp)
 
     # Find index of median prediction for each sample
     weight_cdf = stable_cumsum(sorted_weights, axis=0)
     adjusted_percentile = percentile / 100 * weight_cdf[-1]
 
     # For percentile=0, ignore leading observations with sample_weight=0. GH20528
     mask = adjusted_percentile == 0
-    adjusted_percentile[mask] = np.nextafter(
+    adjusted_percentile[mask] = xp.nextafter(
         adjusted_percentile[mask], adjusted_percentile[mask] + 1
     )
 
-    percentile_idx = np.array(
+    percentile_idx = xp.array(
         [
-            np.searchsorted(weight_cdf[:, i], adjusted_percentile[i])
+            xp.searchsorted(weight_cdf[:, i], adjusted_percentile[i])
             for i in range(weight_cdf.shape[1])
         ]
     )
-    percentile_idx = np.array(percentile_idx)
+    percentile_idx = xp.array(percentile_idx)
     # In rare cases, percentile_idx equals to sorted_idx.shape[0]
     max_idx = sorted_idx.shape[0] - 1
-    percentile_idx = np.apply_along_axis(
-        lambda x: np.clip(x, 0, max_idx), axis=0, arr=percentile_idx
+    percentile_idx = xp.apply_along_axis(
+        lambda x: xp.clip(x, 0, max_idx), axis=0, arr=percentile_idx
     )
 
-    col_index = np.arange(array.shape[1])
+    col_index = xp.arange(array.shape[1])
     percentile_in_sorted = sorted_idx[percentile_idx, col_index]
+    array = xp.asarray(array)
     percentile = array[percentile_in_sorted, col_index]
     return percentile[0] if n_dim == 1 else percentile
+
+
+def _take_along_axis(sample_weight, sorted_idx, xp=None):
```

Comment:
What do you think of moving it to `_array_api.py`? We have been adding things there that we need for array API but that doesn't exist in `xp`. Some of the things there also have restrictions on what they can do (like assuming `axis=0` here). A reason to keep it here is that (after a quick search) it seems `take_along_axis` is only used in `stats.py` and no where else. A reason to put it in `_array_api.py` is that this is where we collect other such customisations.

Overall I have no opinion, I think. But I thought I'd bring it up as something to ponder.